knimin/lib/tests/test_data_access.py

```diff
@@ -1451,6 +1457,60 @@ def _basic_test_steps_for_normalized_shotgun_plate(self, before, after,
 
             self.assertEqual(obs[k], exp_rnsp[k])
 
+    def test_generate_i5_i7_indexes(self):
+        # these tests will depend on what's in the db, which are added
+        # just below -- Add shotgun_index valid values
+
+        # generate 10 and check the values are just fine
+        obs = db.generate_i5_i7_indexes('iTru', 10)
+        exp = [1L, 2L, 3L, 4L, 5L, 6L, 7L, 8L, 9L, 10L]
+        self.assertEqual(obs, exp)
+
+        # ask for another 10 and make sure it starts where it should be
+        obs = db.generate_i5_i7_indexes('iTru', 10)
+        exp = [11L, 12L, 13L, 14L, 15L, 16L, 17L, 18L, 19L, 20L]
+        self.assertEqual(obs, exp)
+
+        # test that we reset index correctly
+        # iTru is 3 and we have 1360 valid barcodes
+        sql = """UPDATE pm.shotgun_index_tech
+                 SET last_index_idx = 1358
+                 WHERE shotgun_index_tech_id = 3"""
+        db._con.execute(sql)
+
+        obs = db.generate_i5_i7_indexes('iTru', 10)
+        exp = [1359L, 1360L, 1L, 2L, 3L, 4L, 5L, 6L, 7L, 8L]
+        self.assertEqual(obs, exp)
+
+        # testing errors
+
+        with self.assertRaises(ValueError) as ctx:
+            db.generate_i5_i7_indexes('Should Error', 10)
+        self.assertEqual(
+            ctx.exception.message, "Not a valid shotgun index technology: "
+            "Should Error")
+
+        with self.assertRaises(ValueError) as ctx:
+            db.generate_i5_i7_indexes('BiooNEXTflex-HT', 10)
+        self.assertEqual(ctx.exception.message,
+                         "'BiooNEXTflex-HT' doesn't have any barcodes")
+
+    def test_get_shotgun_index_information(self):
+        obs = db.get_shotgun_index_information(100)
+        exp = {'shotgun_index_id': 100L, 'i7_row': 3, 'name': 'iTru',
+               'i7_bases': 'TCGGTTAC', 'i7_name': 'iTru7_109_04',
+               'i5_i7_sameplate': False, 'last_index_idx': 0,
+               'dual_index': True, 'i5_name': 'iTru5_01_F', 'i7_col': 21,
+               'i5_row': 5, 'i5_bases': 'GTGAGACT', 'i5_col': 0}
+        self.assertEqual(obs, exp)
+
+        # testing errors
+
+        with self.assertRaises(ValueError) as ctx:
+            db.get_shotgun_index_information(100000000000000)
+        self.assertEqual(ctx.exception.message,
```

Comment:
Just a general FYI, using the context variable is a little bit awkward as a context describes a scope and so this is relying on the context "leaking" the variable. I'm not requesting a change, just noting that this is similar to using a leaked forloop variable.